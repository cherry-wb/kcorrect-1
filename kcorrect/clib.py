# This file was automatically generated by SWIG (http://www.swig.org).
# Version 2.0.4
#
# Do not make changes to this file unless you know what you are doing--modify
# the SWIG interface file instead.


"""
Python interface of Michael Blanton's kcorrect C library

This module provides the Pythonic interface to Michael Blanton's
kcorrect C libray functions.  It is generated by SWIG with the
interface written by Taro Sato.
"""


from sys import version_info
if version_info >= (2,6,0):
    def swig_import_helper():
        from os.path import dirname
        import imp
        fp = None
        try:
            fp, pathname, description = imp.find_module('_clib', [dirname(__file__)])
        except ImportError:
            import _clib
            return _clib
        if fp is not None:
            try:
                _mod = imp.load_module('_clib', fp, pathname, description)
            finally:
                fp.close()
            return _mod
    _clib = swig_import_helper()
    del swig_import_helper
else:
    import _clib
del version_info
try:
    _swig_property = property
except NameError:
    pass # Python < 2.2 doesn't have 'property'.
def _swig_setattr_nondynamic(self,class_type,name,value,static=1):
    if (name == "thisown"): return self.this.own(value)
    if (name == "this"):
        if type(value).__name__ == 'SwigPyObject':
            self.__dict__[name] = value
            return
    method = class_type.__swig_setmethods__.get(name,None)
    if method: return method(self,value)
    if (not static):
        self.__dict__[name] = value
    else:
        raise AttributeError("You cannot add attributes to %s" % self)

def _swig_setattr(self,class_type,name,value):
    return _swig_setattr_nondynamic(self,class_type,name,value,0)

def _swig_getattr(self,class_type,name):
    if (name == "thisown"): return self.this.own()
    method = class_type.__swig_getmethods__.get(name,None)
    if method: return method(self)
    raise AttributeError(name)

def _swig_repr(self):
    try: strthis = "proxy of " + self.this.__repr__()
    except: strthis = ""
    return "<%s.%s; %s >" % (self.__class__.__module__, self.__class__.__name__, strthis,)

try:
    _object = object
    _newclass = 1
except AttributeError:
    class _object : pass
    _newclass = 0


class ARRAY(_object):
    __swig_setmethods__ = {}
    __setattr__ = lambda self, name, value: _swig_setattr(self, ARRAY, name, value)
    __swig_getmethods__ = {}
    __getattr__ = lambda self, name: _swig_getattr(self, ARRAY, name)
    __repr__ = _swig_repr
    __swig_setmethods__["data"] = _clib.ARRAY_data_set
    __swig_getmethods__["data"] = _clib.ARRAY_data_get
    if _newclass:data = _swig_property(_clib.ARRAY_data_get, _clib.ARRAY_data_set)
    __swig_setmethods__["nd"] = _clib.ARRAY_nd_set
    __swig_getmethods__["nd"] = _clib.ARRAY_nd_get
    if _newclass:nd = _swig_property(_clib.ARRAY_nd_get, _clib.ARRAY_nd_set)
    __swig_setmethods__["dims"] = _clib.ARRAY_dims_set
    __swig_getmethods__["dims"] = _clib.ARRAY_dims_get
    if _newclass:dims = _swig_property(_clib.ARRAY_dims_get, _clib.ARRAY_dims_set)
    def __init__(self): 
        this = _clib.new_ARRAY()
        try: self.this.append(this)
        except: self.this = this
    __swig_destroy__ = _clib.delete_ARRAY
    __del__ = lambda self : None;
ARRAY_swigregister = _clib.ARRAY_swigregister
ARRAY_swigregister(ARRAY)

class FILTER_CURVES(_object):
    __swig_setmethods__ = {}
    __setattr__ = lambda self, name, value: _swig_setattr(self, FILTER_CURVES, name, value)
    __swig_getmethods__ = {}
    __getattr__ = lambda self, name: _swig_getattr(self, FILTER_CURVES, name)
    __repr__ = _swig_repr
    __swig_setmethods__["nk"] = _clib.FILTER_CURVES_nk_set
    __swig_getmethods__["nk"] = _clib.FILTER_CURVES_nk_get
    if _newclass:nk = _swig_property(_clib.FILTER_CURVES_nk_get, _clib.FILTER_CURVES_nk_set)
    __swig_setmethods__["filter_n"] = _clib.FILTER_CURVES_filter_n_set
    __swig_getmethods__["filter_n"] = _clib.FILTER_CURVES_filter_n_get
    if _newclass:filter_n = _swig_property(_clib.FILTER_CURVES_filter_n_get, _clib.FILTER_CURVES_filter_n_set)
    __swig_setmethods__["filter_lambda"] = _clib.FILTER_CURVES_filter_lambda_set
    __swig_getmethods__["filter_lambda"] = _clib.FILTER_CURVES_filter_lambda_get
    if _newclass:filter_lambda = _swig_property(_clib.FILTER_CURVES_filter_lambda_get, _clib.FILTER_CURVES_filter_lambda_set)
    __swig_setmethods__["filter_pass"] = _clib.FILTER_CURVES_filter_pass_set
    __swig_getmethods__["filter_pass"] = _clib.FILTER_CURVES_filter_pass_get
    if _newclass:filter_pass = _swig_property(_clib.FILTER_CURVES_filter_pass_get, _clib.FILTER_CURVES_filter_pass_set)
    __swig_setmethods__["maxn"] = _clib.FILTER_CURVES_maxn_set
    __swig_getmethods__["maxn"] = _clib.FILTER_CURVES_maxn_get
    if _newclass:maxn = _swig_property(_clib.FILTER_CURVES_maxn_get, _clib.FILTER_CURVES_maxn_set)
    def __init__(self): 
        this = _clib.new_FILTER_CURVES()
        try: self.this.append(this)
        except: self.this = this
    __swig_destroy__ = _clib.delete_FILTER_CURVES
    __del__ = lambda self : None;
FILTER_CURVES_swigregister = _clib.FILTER_CURVES_swigregister
FILTER_CURVES_swigregister(FILTER_CURVES)


def k_projection_table(*args):
  """
    k_projection_table(fcrv,lambda,vmatrix,zvals,band_shift)

    DESCRIPTION

      Create the rmatrix, a lookup table which speeds analysis

    INPUT

      TO BE WRITTEN

    """
  return _clib.k_projection_table(*args)

def k_reconstruct_maggies(*args):
  """
    k_reconstruct_maggies(coeffs,redshift,zvals,rmatrix)

    DESCRIPTION

      Interpolate the rmatrix

    INPUT

      TO BE WRITTEN

    """
  return _clib.k_reconstruct_maggies(*args)

def k_fit_nonneg(*args):
  """
    k_fit_nonneg(maggies,maggies_ivar,redshift,zvals,rmatrix,maxiter,tolerange,verbose)

    DESCRIPTION

      Nonnegative solutions

    INPUT

      TO BE WRITTEN

    """
  return _clib.k_fit_nonneg(*args)

def k_fit_photoz(*args):
  """
    k_fit_photoz(maggies,maggies_ivar,redshift,zvals,rmatrix,maxiter,tolerange,verbose)

    DESCRIPTION

      Get photometric redshifts

    INPUT

      TO BE WRITTEN

    """
  return _clib.k_fit_photoz(*args)

def k_read_ascii_table(*args):
  """
    k_read_ascii_table(fname)

    DESCRIPTION

      Read an ASCII file in the following format, which is:
     
        <ndim> <size_{0}> ... <size_{ndim-1}>
        <entry_0>
        <entry_1>
        ...
        <entry_n>
        ...
        <entry_{size_0*size_1*..*size_{ndim-1}-1>
     
      where the table element [k,j,i] (for ndim==3) would be the entry
      element n, where n=i*size_2*size1+j*size2+k.

    INPUT

      fname -- Path to an ASCII file

    """
  return _clib.k_read_ascii_table(*args)
TOL = _clib.TOL

def _check_omegas(*args):
  return _clib._check_omegas(*args)
_check_omegas = _clib._check_omegas

def ztor(*args):
  """
    ztor(z, cosmo)

    DESCRIPTION

      Converts redshift z into comoving distance r in km.

    INPUT

      z -- Redshift
      cosmo -- Cosmology given as tuple of the form (omega0, omegal0)

    """
  return _clib.ztor(*args)

def ztodV(*args):
  """
    ztodV(z, cosmo)

    DESCRIPTION

      Converts redshift z to comoving volume element dV in h^-3 Mpc^3.

    INPUT

      z -- Redshift
      cosmo -- Cosmology given as tuple of the form (omega0, omegal0)

    """
  return _clib.ztodV(*args)

def ztov(*args):
  """
    ztov(z, cosmo)

    Convert redshift z to comoving enclosed volume in h^-3 Mpc^3.

    Parameters
    ----------
    z : float
        A redshift value.
    cosmo : (Omega_matter, Omega_Lambda)
        Cosmology given as a tuple.

    """
  return _clib.ztov(*args)

def Vtoz(*args):
  """
    Vtoz(V, cosmo)

    DESCRIPTION

      Converts comoving enclosed volume in h^-3 Mpc^3 to redshift z.

    INPUT

      V -- Comoving enclosed volume in h^-3 Mpc^3
      cosmo -- Cosmology given as tuple of the form (omega0, omegal0)

    """
  return _clib.Vtoz(*args)

def rtoz(*args):
  """
    rtoz(r, cosmo)

    DESCRIPTION

      Converts comoving distance r in km to redshift z.

    INPUT

      r -- Comoving distance in km
      cosmo -- Cosmology given as tuple of the form (omega0, omegal0)

    """
  return _clib.rtoz(*args)

def ztodm(*args):
  """
    ztodm(z, cosmo)

    Convert a redshift z into the distance modulus dm.

    Parameters
    ----------
    z : float
        A redshift value.
    cosmo : (Omega_matter, Omega_Lambda)
        Cosmology given as a tuple.

    """
  return _clib.ztodm(*args)

def dm2z(*args):
  """
    dm2z(dm, cosmo)

    DESCRIPTION

      Converts distance modulus dm into redshift z.

    INPUT

      dm -- Distance measure
      cosmo -- Cosmology given as tuple of the form (omega0, omegal0)

    """
  return _clib.dm2z(*args)

def z2add(*args):
  """
    z2add(z, cosmo)

    DESCRIPTION

      Gives the angular diameter distance in km/s at redshift z.

    INPUT

      z -- Redshift
      cosmo -- Cosmology given as tuple of the form (omega0, omegal0)

    """
  return _clib.z2add(*args)

def z2t(*args):
  """
    z2t(z, cosmo)

    DESCRIPTION

      Returns the age of universe at redshift z in h^-1 Gyr.

    INPUT

      z -- Redshift
      cosmo -- Cosmology given as tuple of the form (omega0, omegal0)

    """
  return _clib.z2t(*args)

def t2z(*args):
  """
    t2z(t, cosmo)

    DESCRIPTION

      Returns the redshift z given the age of universe in h^-1 Gyr.

    INPUT

      t -- Age of universe in h^-1 Gyr
      cosmo -- Cosmology given as tuple of the form (omega0, omegal0)

    """
  return _clib.t2z(*args)
# This file is compatible with both classic and new-style classes.


