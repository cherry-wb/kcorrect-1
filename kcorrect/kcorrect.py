#!/usr/bin/env python2.6
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
#from __future__ import unicode_literals
import numpy as np
from kcorrect.photo import Photo
from kcorrect.utils.cosmology import ztodm
from kcorrect.utils.photo import maggie2ab, fit_nonneg, reconstruct_maggie
from kcorrect.globaldefs import FTYPE


__all__ = ['KCorrect', 'KCorrectAB']


class KCorrect(Photo):
    """
    k correction class for photometry in maggies

    This is the base class for k correction which implements the most
    basic functionalities.  AB maggies are handy to deal with
    internally, but they are not really the conventional photometric
    system observers are used to.  Therefore one usually wants to
    subclass KCorrect to implement a conversion from the input
    photometry to AB maggies by overriding set_data and convert2maggie
    methods.
    """

    def __init__(self, *args, **kwargs):
        """
        Parameters
        ----------
        ptable : ProjectionTableDB object
            Defaults to ProjectionTableDB().  If a non-standard
            projection table (i.e., with different redshift intervals
            or template spectra, etc.) is to be used in reconstructing
            AB maggies, then a custom ProjectionTableDB object can be
            specified here.
        cosmo : (Omega_matter, Omega_Lambda, h_100)
        maxiter : int
            Maximum number of iterations for non-negative fitting.
        verbosity : int
            Set the level of verbosity for output; currently not
            really used.
        """
        self.maxiter = kwargs['maxiter'] if 'maxiter' in kwargs else 3000
        super(KCorrect, self).__init__(*args, **kwargs)
        self._model_maggie = None
        self._distmod = None

    def set_input(self, redshift, filter_list, maggie, maggie_ivar,
                  *args, **kwargs):
        """
        Set input data

        INPUT

        redshift -- List of redshifts
        filter_list -- A FilterList object
        maggie -- List of AB maggies
        maggie_ivar -- List of inverse variances of AB maggies
        """
        super(KCorrect, self).set_input_photo(filter_list, maggie, maggie_ivar)

        redshift = np.atleast_1d(np.asarray(redshift, dtype=FTYPE))
        self.redshift = redshift.reshape((len(self), 1))

        # zero redshifts for convenience
        self.redshift0 = np.zeros(self.redshift.shape, dtype=FTYPE)

        # get fit coefficients once and for all
        self.do_fit(maxiter=self.maxiter)

    @property
    def model_maggie(self):
        """
        Maggies of the model SED through the input passbands.
        
        model_maggie only differs from input_maggie in that they are
        computed from the best fit SEDs through the input passbands;
        these two maggies should be very close to each other if the
        fits are good.
        """
        if self._model_maggie is not None:
            return self._model_maggie
        self._model_maggie = reconstruct_maggie(self.coeffs, self.redshift,
                                                self.ptable[self.filter_list])
        return self._model_maggie

    @property
    def distmod(self):
        """
        Distance moduli to input redshifts.
        """
        if self._distmod is not None:
            return self._distmod
        dm = ztodm(self.redshift, self.cosmo)
        self._distmod = np.asarray(dm, dtype=FTYPE)
        return self._distmod

    def do_fit(self, maxiter=3000):
        """
        Do non-negative fitting.
        """
        self._model_maggie = None
        r = fit_nonneg(self.input_maggie, self.input_maggie_ivar,
                       self.redshift, self.ptable[self.filter_list],
                       maxiter)
        self.coeffs, self.chi2, self.iter = r

    def kcorrect(self, filter_list_q=None, filter_list_r=None, band_shift=0.):
        """
        Compute the k-corrections between two sets of filters

        This method computes k corrections bewteen two sets of filters
        such that

        m_r = M_q + DM(z) + K_qr(z)

        where m_r is the apparent magnutude through filter r, M_q the
        absolute magnitude through filter q, DM the distance modulus,
        and K_qr is the k correction between filters q and r.

        The number of filters in sets q and r must match.

        INPUT

        filter_list_q, filter_list_r
            -- FilterList objects; if not provided explicitly, q and r
               both defaults to self.filter_list

        band_shift -- Amount of redshift used for computing the
                      absolute magnitudes through band-shifted
                      filters; defaults to zero.
        """
        if not filter_list_q: filter_list_q = self.filter_list
        if not filter_list_r: filter_list_r = self.filter_list

        if len(filter_list_q) != len(filter_list_r):
            raise ValueError('Numbers of filters must match.')

        if filter_list_r != self.filter_list:
            rm2 = reconstruct_maggie(self.coeffs, self.redshift,
                                     self.ptable[filter_list_r])
        else:
            rm2 = self.model_maggie

        if band_shift > 0.:
            bs = np.ones(self.redshift.shape, dtype=FTYPE) * band_shift
            rm1 = reconstruct_maggie(self.coeffs, bs,
                                     self.ptable[filter_list_q])
            rm1 = rm1 / (1. + bs)
        else:
            rm1 = reconstruct_maggie(self.coeffs, self.redshift0,
                                     self.ptable[filter_list_q])
        return 2.5 * np.log10(rm1 / rm2)

    def absmag(self, filter_list=None, band_shift=0.):
        """
        Compute the absolute AB magnitude through a set of filters

        This method returns the absolute AB magnitudes through a set
        of (band-shifted) filters, which defaults to the input
        filters.  The output magnitudes are computed from the input
        photometries where available; otherwise the output magnitudes
        are computed from the model photometries reconstructed from
        the best fit SED.

        INPUT

        filter_list -- A FilterList object; defaults to the input filters
        band_shift -- Amount of redshift used for computing the
                      absolute magnitudes through band-shifted
                      filters; defaults to zero.
        """
        if not filter_list:
            filter_list = self.filter_list

        # get model maggies for abs mag
        bs = np.ones(self.redshift.shape, dtype=FTYPE) * band_shift
        rm = reconstruct_maggie(self.coeffs, bs, self.ptable[filter_list])
        rm = rm / (1. + bs)

        # fill all entries with model
        mag = maggie2ab(rm) - self.distmod
        if len(filter_list) == len(self.filter_list):
            # now replace those entries with good input and well
            # defined ivar
            mask = (np.greater(self.input_maggie,0.) *
                    np.greater(self.input_maggie_ivar, 0.)).astype(int)
            kcorr = self.kcorrect(filter_list_q=filter_list,
                                  band_shift=band_shift)
            mag = ((1 - mask) * mag
                   + mask * (maggie2ab(self.input_maggie)
                             - self.distmod - kcorr))
        return mag

    def appmag(self, filter_list, redshift=[]):
        """
        Compute apparent AB magnitudes through a set of filters

        Computes the apparent magnitudes through a set of filters.
        Optionally, the objects to be observed can be moved to
        different redshifts than the input redshifts; i.e., the best
        fit SEDs are redshifted arbitrarily and *observed* with the
        set of filters.

        INPUT

        filter_list -- A FilterList object
        redshift -- A redshift or a 1D array of redshift (with the
                    number of elements equal to the number of input
                    objects) to observe galaxies at arbitrary
                    redshifts
        """
        redshift = np.atleast_1d(redshift).astype(FTYPE)
        if redshift.size == 0:
            # simply observe the input objects with given filters
            maggie = reconstruct_maggie(self.coeffs, self.redshift,
                                        self.ptable[filter_list])
            return maggie2ab(maggie)

        # when redshifts are given, use them
        if redshift.size == 1:
            redshift = np.ones(self.redshift.shape, dtype=FTYPE) * redshift[0]
        elif redshift.size == self.redshift.size:
            redshift = np.reshape(redshift, self.redshift.shape)
        else:
            raise ValueError('Inconsistent redshift array')
        maggie_S = reconstruct_maggie(self.coeffs, redshift,
                                      self.ptable[filter_list])
        m_S = maggie2ab(maggie_S)
        dm1 = ztodm(self.redshift, self.cosmo)
        dm2 = ztodm(redshift, self.cosmo)

        # because of what reconstruct_maggie returns, the flux needs
        # to be rescaled by the ratio of distances
        return m_S - dm1 + dm2

    def model_spectrum(self, index=None, lmin=None, lmax=None):
        """
        Returns the best fit SED(s) in the rest frame.

        The output SED is in units of erg s^-1 cm^-2 A^-1.

        Parameters
        ----------
        index : array of indices
            The index(ices) of the object for which spectrum is obtained.
        lmin : float
            Minimum wavelength.
        lmax : float
            Maximum wavelength.
        """
        index = (np.arange(0, len(self)) if index is None
                 else np.atleast_1d(np.asarray(index, int)))

        lamb, vmatrix = self.ptable[self.filter_list].templates
        lamb = 0.5 * (lamb[:-1] + lamb[1:])

        mask = np.ones(lamb.shape, dtype=bool)
        if lmin is not None:
            mask = mask * np.greater(lamb, lmin)
        if lmax is not None:
            mask = mask * np.less(lamb, lmax)
        lamb = lamb[mask]

        vmatrix = np.compress(mask, np.transpose(vmatrix), axis=0)

        data = []
        for coeffs in self.coeffs[index]:
            flux = np.add.reduce(np.transpose(vmatrix * coeffs))
            data.append(flux)
        return (lamb, data if len(data) > 1 else data[0])


class KCorrectAB(KCorrect):
    """
    K-correction for AB photometry

    Expect AB magnitudes and 1-sigma uncertainties
    """

    def set_input(self, filter_list, abmag, abmag_sigma, redshift,
                  *args, **kwargs):
        """
        Set input data

        If abmag_sigma is not given, the minimum uncertainties of 0.02
        mag will be assumed.

        INPUT

        redshift -- List of redshifts
        filter_list -- FilterList object
        abmag -- List of AB magnitudes
        abmag_sigma -- List of 1-sigma uncertainties in AB magnitudes;
                       defaults to an empty list
        """

        # if uncertainties not given, assume 2% errors
        abmag_sigma = (np.ones(abmag.shape, dtype=FTYPE) * 0.02
                       if abmag_sigma is None else abmag_sigma)

        # KCorrect.set_data expects maggies
        maggie, maggie_sigma = ab2maggie(abmag, abmag_sigma)
        maggie, maggie_ivar = maggie, sigma2ivar(maggie_sigma)

        super(KCorrect, self).set_input(filter_list, maggie, maggie_ivar,
                                        redshift, *args, **kwargs)
