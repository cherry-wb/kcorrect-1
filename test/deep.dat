# redshift redshift_err B R I
1.029e+00 1.465e-05 23.112 22.363 21.690
9.581e-01 6.284e-05 24.668 23.878 23.248
1.219e+00 8.371e-05 24.847 23.551 22.720
1.025e+00 2.837e-05 24.217 23.416 22.689
1.115e+00 9.263e-05 25.398 22.802 21.532
1.098e+00 3.097e-05 23.731 23.836 23.518
1.199e+00 6.882e-05 22.782 22.680 22.356
8.771e-01 1.686e-05 24.142 23.969 23.327
4.574e-01 1.074e-04 23.702 23.721 23.675
1.081e+00 1.863e-04 24.778 23.740 23.110
