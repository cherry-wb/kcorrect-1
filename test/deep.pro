openr, lun, 'deep.dat', /GET_LUN
header = STRARR(1)
READF, lun, header

rows = 10
data = FLTARR(5, rows)  
READF, lun, data

redshift = data(0,*)
mag = data(2:4,*)

; prepare zcat like structure
zcatel = create_struct('magb',0.,'magr',0.,'magi',0.)
zcat = replicate(zcatel,10)
zcat.magb = reform(mag[0,*])
zcat.magr = reform(mag[1,*])
zcat.magi = reform(mag[2,*])

kcorr_nub = fltarr(3,10)
kcorr_nub = deep_kcorrect(redshift,zcat=zcat,absmag=absmag_nub, $
                          omaggies=om,oivar=oi)


fl_deep = 'deep_'+['B','R','I']+'.par'
ngals = n_elements(redshift)

kcorrect, om,oi, redshift, kcorr_deep, vmatrix=vmatrix, lambda=lambda, $
  coeffs=coeffs, rmatrix=rmatrix, zvals=zvals, filterlist=fl_deep, $
  absmag = absmag_deep

redshift0 = replicate(0.,n_elements(redshift))
distmod = lf_distmod(redshift)
distmod0 = lf_distmod(redshift0)

print, '### k-corrections: DEEP BRI ###'
print, kcorr_deep
print, ''

print, '### absolute magnitudes: DEEP BRI ###'
print, absmag_deep
print, ''

band_shift = 1.0

kcorrect, om,oi, redshift, kcorr_deep, $
  filterlist=fl_deep, $
  absmag = absmag_deep, band_shift = band_shift

print, '### k-corrections: DEEP BRI band-shifted to z=1 ###'
print, kcorr_deep
print, ''

print, '### absolute magnitudes: DEEP BRI band-shifted to z=1 ###'
print, absmag_deep
print, ''

print, '### k-corrections: NUB (IDL default) ###'
print, kcorr_nub
print, ''

print, '### absolute magnitudes: NUB (IDL default) ###'
print, absmag_nub
print, ''


fl_bessell = 'bessell_'+['U','B','V','R','I']+'.par'
redshift0 = replicate(0.,n_elements(redshift))
distmod = lf_distmod(redshift)
distmod0 = lf_distmod(redshift0)

appmag = fltarr(n_elements(fl_bessell), ngals)
k_reconstruct_maggies, coeffs, redshift0, rm0, $
  vmatrix=vmatrix, lambda=lambda, filterlist=fl_bessell
k_reconstruct_maggies, coeffs, redshift, rm1, $
  vmatrix=vmatrix, lambda=lambda, filterlist=fl_bessell

kcorr = 2.5*alog10(rm0/rm1)

appmag = -2.5*alog10(rm0)

vega2ab = k_vega2ab(filterlist=fl_bessell,/kurucz)
vegamag = fltarr(n_elements(fl_bessell), ngals)
for i = 0L, n_elements(fl_bessell)-1L do $
  vegamag[i,*] = appmag[i,*] -vega2ab[i] - distmod; - kcorr[i]
print, '### absolute magnitudes: Bessell UBVRI on Vega scale ###'
print, vegamag
print, ''

fl_sdss = 'sdss_'+['u','g','r','i','z']+'0.par'

redshift01 = replicate(0.1,ngals)
distmod01 = lf_distmod(redshift01)
k_reconstruct_maggies, coeffs, redshift01, rm, $
  vmatrix=vmatrix, lambda=lambda, filterlist=fl_sdss
appmag = -2.5*alog10(rm)

for i = 0L, n_elements(fl_bessell)-1L do $
  appmag[i,*] = appmag[i,*] - distmod + distmod01

print, '### apparent magnitude SDSS ugriz if at z = 0.1 ###'
print, appmag
print, ''










exit



kcorrect, om,oi, redshift, kcdm, vmatrix=vmatrix, lambda=lambda, $
  coeffs=coeffs, rmatrix=rmatrix, zvals=zvals

ngals = n_elements(redshift)

redshift0 = replicate(0.,n_elements(redshift))
distmod = lf_distmod(redshift)
distmod0 = lf_distmod(redshift0)

fl_sdss = 'sdss_'+['u','g','r','i','z']+'0.par'
fl = 'bessell_'+['U','B','V','R','I']+'.par'

k_reconstruct_maggies, coeffs, redshift0, rm0, $
  vmatrix=vmatrix, lambda=lambda, filterlist=fl
k_reconstruct_maggies, coeffs, redshift, rm1, $
  vmatrix=vmatrix, lambda=lambda, filterlist=fl_sdss

kcorr = 2.5*alog10(rm0/rm1)

abmag = fltarr(n_elements(fl),ngals)
