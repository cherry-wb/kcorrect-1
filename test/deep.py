#!/usr/bin/env python2.6
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
#from __future__ import unicode_literals
import numpy as np
import kcorrect as KC
from kcorrect.deep import DEEPKCorrect, DEEPPhotoZ


def main():
    d = np.loadtxt('test/deep.dat',
                   dtype=[('z', float), ('dz', float),
                          ('B', float), ('R', float), ('I', float)])
    z = d['z']
    mag = np.transpose([d['B'], d['R'], d['I']])

    # testing photo-z
    pz = DEEPPhotoZ(mag)
    print(pz.redshift)
    print(z)
    exit()
    # photo-z end


    kc = DEEPKCorrect(z, mag, cosmo=(0.3, 0.7, 1.0))

    print('### k-corrections: DEEP BRI ###')
    kcorr = kc.kcorrect()
    for each in kcorr:
        print('%+.6f %+.6f %+.6f' % tuple(each))
    print()

    print('### absolute magnitudes: DEEP BRI ###')
    absmag = kc.absmag()
    for each in absmag:
        print('%+.4f %+.4f %+.4f' % tuple(each))
    print()

    print('### k-corrections: DEEP BRI band-shifted to z=1 ###')
    band_shift = 1.0
    kcorr = kc.kcorrect(band_shift=band_shift)
    for each in kcorr:
        print('%+.6f %+.6f %+.6f' % tuple(each))
    print()

    print('### absolute magnitudes: DEEP BRI band-shifted to z=1 ###')
    absmag = kc.absmag(band_shift=band_shift)
    for each in absmag:
        print('%+.4f %+.4f %+.4f' % tuple(each))
    print()

    print('### k-corrections: NUB (IDL default) ###')
    fl = KC.FilterList(['galex_NUV', 'bessell_U', 'bessell_B'])
    kcorr = kc.kcorrect(fl)
    for each in kcorr:
        print('%+.6f %+.6f %+.6f' % tuple(each))
    print()

    print('### absolute magnitudes: NUB (IDL default) ###')
    absmag = kc.absmag(fl)
    for each in absmag:
        print('%+.4f %+.4f %+.4f' % tuple(each))
    print()

    print('### absolute magnitudes: Bessell UBVRI on Vega scale ###')
    fl = KC.FilterList(['bessell_U', 'bessell_B', 'bessell_V',
                        'bessell_R', 'bessell_I'])
    vega2ab = KC.vega2ab(fl)
    absmag = kc.absmag(fl) - vega2ab
    for each in absmag:
        print('%+.4f %+.4f %+.4f %+.4f %+.4f' % tuple(each))
    print()

    print('### apparent magnitude SDSS ugriz if at z = 0.1 ###')
    fl = KC.FilterList(['sdss_u0', 'sdss_g0', 'sdss_r0',
                        'sdss_i0', 'sdss_z0'])
    mag = kc.appmag(fl,redshift=0.1)
    for each in mag:
        print('%+.4f %+.4f %+.4f %+.4f %+.4f' % tuple(each))
    print()


if __name__ == '__main__':
    main()
