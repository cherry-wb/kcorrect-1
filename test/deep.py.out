### k-corrections: DEEP BRI ###
+0.914720 +0.651905 +0.146177
+0.812483 +0.310862 -0.246056
+2.138424 +1.774332 +1.327768
+1.038617 +0.744096 +0.189650
+3.920245 +2.525232 +1.697521
-0.271478 +0.876878 +1.078303
-0.007348 +0.462779 +0.401534
+0.169872 +1.005893 +0.848081
-0.390317 -0.167881 +0.127659
+1.216219 +0.580401 +0.078876

### absolute magnitudes: DEEP BRI ###
-21.2052 -21.6914 -21.8587
-19.3553 -19.6437 -19.7168
-21.1492 -22.0811 -22.4655
-20.2136 -20.7201 -20.8927
-22.1403 -23.3413 -23.7836
-19.5743 -20.6177 -21.1371
-21.0239 -21.5961 -21.8588
-19.0020 -20.0110 -20.4952
-17.1631 -17.3665 -17.7081
-19.9731 -20.3753 -20.5038

### k-corrections: DEEP BRI band-shifted to z=1 ###
-0.762833 -0.748914 -0.738288
-0.762894 -0.753150 -0.789595
-0.664695 -0.541351 -0.341209
-0.763494 -0.748957 -0.728969
-0.696778 -0.509272 -0.436032
-0.825495 -0.789694 -0.689832
-0.848051 -0.782977 -0.617951
-0.650186 -0.782598 -0.884224
-0.238430 -0.508308 -0.431648
-0.740974 -0.741275 -0.670759

### absolute magnitudes: DEEP BRI band-shifted to z=1 ###
-19.5276 -20.2906 -20.9742
-17.7800 -18.5797 -19.1733
-18.3460 -19.7654 -20.7965
-18.4115 -19.2271 -19.9741
-17.5233 -20.3068 -21.6501
-19.0203 -18.9511 -19.3690
-20.1832 -20.3503 -20.8393
-18.1819 -18.2225 -18.7629
-17.3150 -17.0261 -17.1487
-18.0159 -19.0536 -19.7541

### k-corrections: NUB (IDL default) ###
-0.692655 -0.541377 -0.555010
-0.673314 -0.576753 -0.646636
-0.520448 -0.212312 -0.057172
-0.690055 -0.512592 -0.536579
-0.408013 -0.085099 -0.043284
-0.844229 -0.666840 -0.513597
-0.848206 -0.630640 -0.465148
-0.663726 -0.620860 -0.682486
-0.260737 -0.473638 -0.420596
-0.616628 -0.528817 -0.498211

### absolute magnitudes: NUB (IDL default) ###
-19.5978 -20.4981 -21.1575
-17.8695 -18.7561 -19.3162
-18.4903 -20.0944 -21.0806
-18.4850 -19.4634 -20.1664
-17.8121 -20.7310 -22.0428
-19.0016 -19.0740 -19.5452
-20.1831 -20.5026 -20.9921
-18.1684 -18.3842 -18.9646
-17.2927 -17.0608 -17.1598
-18.1403 -19.2661 -19.9267

### absolute magnitudes: Bessell UBVRI on Vega scale ###
-21.2920 -21.0675 -21.4880 -21.8776 -22.3035
-19.5498 -19.2263 -19.5258 -19.8364 -20.1699
-20.8883 -20.9906 -21.6928 -22.2583 -22.8920
-20.2573 -20.0765 -20.4998 -20.9027 -21.3349
-21.5249 -21.9528 -22.8950 -23.5199 -24.2095
-19.8686 -19.4551 -20.0964 -20.7833 -21.5474
-21.2965 -20.9022 -21.3072 -21.7710 -22.2896
-19.1781 -18.8746 -19.5212 -20.1777 -20.9088
-17.9108 -17.0483 -17.1579 -17.5885 -18.1070
-20.0599 -19.8367 -20.2108 -20.5651 -20.9523

### apparent magnitude SDSS ugriz if at z = 0.1 ###
+17.1768 +16.3375 +15.9351 +15.6949 +15.5741
+18.8810 +18.1730 +17.9092 +17.7616 +17.7158
+17.7300 +16.4349 +15.7003 +15.2813 +14.9712
+18.2438 +17.3304 +16.9236 +16.6567 +16.5408
+17.2364 +15.4888 +14.4814 +14.0233 +13.6522
+18.4939 +17.9384 +17.2847 +16.7121 +16.3042
+17.1030 +16.4891 +16.1111 +15.7596 +15.5769
+19.2305 +18.5263 +17.8641 +17.3244 +16.9451
+20.3503 +20.2981 +20.2784 +19.9074 +19.7511
+18.4143 +17.5693 +17.2172 +17.0216 +16.9292

